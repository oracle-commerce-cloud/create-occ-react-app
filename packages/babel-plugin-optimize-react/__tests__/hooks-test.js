'use strict';

const plugin = require('../index.js');
const babel = require('@babel/core');

function transform(code) {
  return babel.transform(code, {
    plugins: [plugin],
  }).code;
}

describe('React hook transforms', () => {
  it('should support transform hook imports', () => {
    const test = `
      import React, {useState} from "react";
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support hook imports with aliasing', () => {
    const test = `
      import React, {useState as foo} from "react";
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support destructuring hooks from imports', () => {
    const test = `
      import React, {useState} from "react";

      export function MyComponent() {
        const [counter, setCounter] = useState(0);

        return React.createElement("div", null, counter);
      }
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support destructuring hooks from imports #2', () => {
    const test = `
      import React from "react";
      const {useState} = React;

      export function MyComponent() {
        const [counter, setCounter] = useState(0);

        return React.createElement("div", null, counter);
      }
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support destructuring hooks from imports #3', () => {
    const test = `
      import React from "react";
      const useState = React.useState;

      export function MyComponent() {
        const [counter, setCounter] = useState(0);

        return React.createElement("div", null, counter);
      }
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support destructuring hooks from imports #4', () => {
    const test = `
      import React from "react";
      const foo = React.useState;

      export function MyComponent() {
        const [counter, setCounter] = foo(0);

        return React.createElement("div", null, counter);
      }
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support destructuring hooks from imports #5', () => {
    const test = `
      import React, {useState as foo} from "react";

      export function MyComponent() {
        const [counter, setCounter] = foo(0);

        return React.createElement("div", null, counter);
      }
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support destructuring hooks from require calls', () => {
    const test = `
      const React = require("react");
      const {useState} = React;

      export function MyComponent() {
        const [counter, setCounter] = useState(0);

        return React.createElement("div", null, counter);
      }
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support hook imports with no default', () => {
    const test = `
      import {useState} from "react";
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support hook CJS require with no default', () => {
    const test = `
      const {useState} = require("react");
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support mixed hook imports', () => {
    const test = `
      import React from "react";
      import {memo, useState} from "react";
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support mixed hook imports with no default', () => {
    const test = `
      import {useState} from "react";
      import {memo} from "react";
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });

  it('should support mixed hook imports with no default #2', () => {
    const test = `
      import {memo, useRef, useState} from "react";

      export const Portal = memo(() => null);
    `;
    const output = transform(test);
    expect(output).toMatchSnapshot();
  });
});
