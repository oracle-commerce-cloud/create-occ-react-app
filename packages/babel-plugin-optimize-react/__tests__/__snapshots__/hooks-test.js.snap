// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React hook transforms should support destructuring hooks from imports #2 1`] = `
"import React from \\"react\\";
const __reactCreateElement__ = React.createElement;
const {
  useState
} = React;
export function MyComponent() {
  const _ref_0 = useState(0);

  const setCounter = _ref_0[1];
  const counter = _ref_0[0];
  return __reactCreateElement__(\\"div\\", null, counter);
}"
`;

exports[`React hook transforms should support destructuring hooks from imports #3 1`] = `
"import React from \\"react\\";
const __reactCreateElement__ = React.createElement;
const useState = React.useState;
export function MyComponent() {
  const _ref_0 = useState(0);

  const setCounter = _ref_0[1];
  const counter = _ref_0[0];
  return __reactCreateElement__(\\"div\\", null, counter);
}"
`;

exports[`React hook transforms should support destructuring hooks from imports #4 1`] = `
"import React from \\"react\\";
const __reactCreateElement__ = React.createElement;
const foo = React.useState;
export function MyComponent() {
  const _ref_0 = foo(0);

  const setCounter = _ref_0[1];
  const counter = _ref_0[0];
  return __reactCreateElement__(\\"div\\", null, counter);
}"
`;

exports[`React hook transforms should support destructuring hooks from imports #5 1`] = `
"import React from \\"react\\";
const __reactCreateElement__ = React.createElement;
const {
  useState: foo
} = React;
export function MyComponent() {
  const _ref_0 = foo(0);

  const setCounter = _ref_0[1];
  const counter = _ref_0[0];
  return __reactCreateElement__(\\"div\\", null, counter);
}"
`;

exports[`React hook transforms should support destructuring hooks from imports 1`] = `
"import React from \\"react\\";
const __reactCreateElement__ = React.createElement;
const {
  useState
} = React;
export function MyComponent() {
  const _ref_0 = useState(0);

  const setCounter = _ref_0[1];
  const counter = _ref_0[0];
  return __reactCreateElement__(\\"div\\", null, counter);
}"
`;

exports[`React hook transforms should support destructuring hooks from require calls 1`] = `
"const React = require(\\"react\\");

const __reactCreateElement__ = React.createElement;
const {
  useState
} = React;
export function MyComponent() {
  const _ref_0 = useState(0);

  const setCounter = _ref_0[1];
  const counter = _ref_0[0];
  return __reactCreateElement__(\\"div\\", null, counter);
}"
`;

exports[`React hook transforms should support hook CJS require with no default 1`] = `
"const {
  useState
} = require(\\"react\\");"
`;

exports[`React hook transforms should support hook imports with aliasing 1`] = `
"import React from \\"react\\";
const {
  useState: foo
} = React;"
`;

exports[`React hook transforms should support hook imports with no default 1`] = `
"import React from \\"react\\";
const {
  useState
} = React;"
`;

exports[`React hook transforms should support mixed hook imports 1`] = `
"import React from \\"react\\";
import \\"react\\";
const {
  memo,
  useState
} = React;"
`;

exports[`React hook transforms should support mixed hook imports with no default #2 1`] = `
"import React from \\"react\\";
const {
  memo,
  useRef,
  useState
} = React;
export const Portal = memo(() => null);"
`;

exports[`React hook transforms should support mixed hook imports with no default 1`] = `
"import React from \\"react\\";
const {
  useState
} = React;
import \\"react\\";
const {
  memo
} = React;"
`;

exports[`React hook transforms should support transform hook imports 1`] = `
"import React from \\"react\\";
const {
  useState
} = React;"
`;
